adrs/0049-prebuilt-rubies-at-runtime.md

```diff
@@ -0,0 +1,75 @@
+# ADR-0049 Prebuilt Rubies at Runtime
+
+***Status***: proposed
+
+## Context
+Currently, only two versions of Ruby are available at runtime as part of the hosted build VM tools cache.  The setup-ruby action only supports using what's in the cache and does not support pulling other versions at runtime.  It also doesn't support pulling other versions like JRuby at runtime.
+
+This is in contrast to other setup actions like setup-node where the cache is an optimization but supports pulling any version by a specified semver at runtime.
+
+## Goals
+
+- Offer as many versions as possible by platforms
+- Offer popular Ruby variants like JRuby
+- Consistent with other setup-xxx actions patterns and Actions workflows.
+
+## Decisions
+
+### Prebuilt Rubies
+
+Other tools like offer a [distribution](https://nodejs.org/dist/) which offers a [queryable endpoint](https://nodejs.org/dist/index.json) which allows the desired version [by a version spec which is semver](https://github.com/actions/setup-node#setup-node).
+
+Other CI services like Travis support pulling a wide variety of [ruby versions offered here](http://rubies.travis-ci.org/).
+
+Requirements:
+- Offer prebuilt Ruby, JRuby, TruffleRuby versions 
+- Setup-Ruby exposes `ruby-version` which accepts a semver
+- Rubies are discoverable and queryable by the Setup-Ruby action.
+- Initially offer for all platforms offered by the hosted Actions VMs (Ubuntu 18, Mac 10.15 and Windows 2019)
+
+NOTE: The prebuilt rubies offered will only be supported for use by GitHub Actions.
+
+The scripts and tooling to build the Rubies will be at `actions/build-ruby` and it will be consumed by the existing action `actions/setup-ruby`.  Since the capabilities are additive, there's no need to create a `v2`.  There is no compat break.
+
+A specific version of a Ruby will be offered as an individual semi-immutable build-ruby repo release similar to how the [actions runner exposes versions](https://github.com/actions/runner/releases/tag/v2.164.0) where each version offers n platforms.  This is also similar to [Travis individual versions](http://rubies.travis-ci.org/ubuntu/18.04/s390x/ruby-2.6.5).
+
+NOTE: semi-immutable means it's desirable to be immutable but it's possible to patch.  Note that 
+
+Option 1: GitHub GPR Universal Package
+
+The `actions/build-ruby` repo offers packages for [example](https://github.com/actions/setup-ruby/packages). 
```

Comment:
What's the difference between packages and releases?

From https://github.com/features/packages#pricing I can see 500MB (total) could quickly become too tight for hosting all Ruby versions. Releases don't have such restrictions AFAIK (just no single file >2GB).

Reply:
I'm discussing with packaging folks.  But essentially, packaging has full semver support without putting that burden on the action or some other index json or api.

I think if we went the packaging route, then each distribution, per version, per arch would be an independent package (how big is a discrete ruby? 10 - 20 MB?)

Reply:
See https://github.com/eregon/ruby-install-builder/releases/tag/builds-bundler1
So it's between 10MB and 150MB. ~35MB for Ruby on Linux.

I think Rubyists are BTW more used to `2.6` rather than `2.6.x`. In use-ruby-action I just use `startsWith` on an ordered list of versions per engine, and it's very simple in the end:
https://github.com/eregon/use-ruby-action/blob/933f49684485836830e31e48a48835cba066dc55/index.js#L77-L79